web/src/user_profile.js

```diff
@@ -96,6 +98,32 @@ export function update_user_profile_streams_list_for_users(user_ids) {
     }
 }
 
+export function update_profile_modal_ui(user, new_data) {
+    if (!(modals.any_active() && modals.active_modal() === "#user-profile-modal")) {
+        return;
+    }
+    if (new_data.bot_owner_id !== undefined) {
+        const $bot_owner_field = $(".bot_owner_user_field");
+        $bot_owner_field.attr("data-field-id", new_data.bot_owner_id);
+    }
+    if (new_data.avatar_url !== undefined) {
+        $("#avatar").css("background-image", `url(${people.medium_avatar_url_for_person(user)})`);
+    }
+    if (new_data.delivery_email !== undefined) {
+        $("email").find(".value").text(new_data.delivery_email);
+    }
+    if (new_data.role !== undefined) {
+        $("#user-type").find(".value").text(new_data.role);
+    }
+    if (new_data.full_name !== undefined || new_data.role !== undefined) {
+        const user_type = {
+            name: user.full_name,
+            should_add_guest_user_indicator: people.should_add_guest_user_indicator(user.user_id),
+        };
+        $("#name .user_profile_name").html(render_user_full_name(user_type));
+    }
+}
```

Comment:
This feels like a lot of duplicated code. Would it be possible to use the `settings_components.ts` system for this diffing logic?

Reply:
Not sure that I am getting what you mean. You are talking about that bunch of `if` statements and updating the profile field values right?
If yes then the answer to your question will be no, the closest a function could be helpful is `set_input_element_value`. Most functions in `settings_components.ts` has to do with getting and setting values input elements. But in our case we are setting text values of div blocks. Also if we are able to make changes in any existing function to set values in div block that would only work for two profile fields (`email` and `role`). Rest of the fields require setting value to different things. Like bot `bot_owner_field` requires updating `data-field-id` attribute of the div block and `avatar` requires updating `background-image` property.

Reply:
I think ultimately we'll rewrite this component using something like Solid and delete this live-update rendering code, so not going to stress about it too much, but it likely would have been simpler to just trigger a full rerender of the modal when values change.

Reply:
Yeah, in the first PR (completion candidate one) the whole modal was being rerender. To optimize that this approach was suggested.